src/lib/kadm5/srv/svr_principal.c

```diff
@@ -1105,45 +1115,59 @@ check_pw_reuse(krb5_context context,
 static
 int create_history_entry(krb5_context context,
                          krb5_keyblock *hist_key, int n_key_data,
-                         krb5_key_data *key_data, osa_pw_hist_ent *hist)
+                         krb5_key_data *key_data, osa_pw_hist_ent *hist_out)
 {
     krb5_error_code ret;
     krb5_keyblock key;
     krb5_keysalt salt;
-    int i;
+    krb5_ui_2 kvno;
+    osa_pw_hist_ent hist;
 
-    hist->key_data = k5calloc(n_key_data, sizeof(krb5_key_data), &ret);
-    if (hist->key_data == NULL)
-        return ret;
+    hist_out->key_data = NULL;
+    hist_out->n_key_data = 0;
 
-    for (i = 0; i < n_key_data; i++) {
-        ret = krb5_dbe_decrypt_key_data(context, NULL, &key_data[i], &key,
+    if (n_key_data <= 0)
+        return EINVAL;
+
+    memset(&key, 0, sizeof(key));
+
+    hist.key_data = k5calloc(n_key_data, sizeof(krb5_key_data), &ret);
+    if (hist.key_data == NULL)
+        goto cleanup;
+
+    /*
+     * We only want to store the most recent kvno, and key_data should already
+     * be sorted in descending order by kvno.
+     */
+    kvno = key_data[0].key_data_kvno;
+
+    for (hist.n_key_data = 0; hist.n_key_data < n_key_data; hist.n_key_data++) {
```

Comment:
I think the loop read a lot better with "i" as the loop index, since it is used six times within the loop.  hist.n_key_data can still be used as the index into hist.key_data, and incremented each time we encrypt a key into the array.